tensorflow_probability/python/sts/trigonometric.py

```diff
@@ -0,0 +1,225 @@
+          name=name)
+
+  @property
+  def drift_scale(self):
+    """Standard deviation of the drift in the cyclic effects."""
+    return self._drift_scale
+
+  @property
+  def observation_noise_scale(self):
+    """Standard deviation of the observation noise."""
+    return self._observation_noise_scale
+
+  @property
+  def period(self):
+    """The seasonal period."""
+    return self._period
+
+  @property
+  def selected_frequencies(self):
+    """Chosen frequency indices to include."""
+    return self._selected_frequencies
+
+
+def build_trigonometric_transition_matrix(period, selected_frequencies, dtype):
+  """Build the transition matrix for a TrigonometricStateSpaceModel"""
+
+  two_pi = tf.constant(2. * np.pi, dtype=dtype)
+  frequencies = two_pi * selected_frequencies / period
+  num_frequencies = len(selected_frequencies)
+
+  sin_frequencies = tf.sin(frequencies)
+  cos_frequencies = tf.cos(frequencies)
+
+  trigonometric_values = tf.stack(
+      [cos_frequencies, sin_frequencies, -sin_frequencies, cos_frequencies],
+      axis=-1)
+
+  transition_matrix = tf.linalg.LinearOperatorBlockDiag(
+      [
+          tf.linalg.LinearOperatorFullMatrix(
+              matrix=tf.reshape(trigonometric_values[i], [2, 2]),
+              is_square=True
+          )
+          for i in range(num_frequencies)
+      ]
+  )
+
+  return transition_matrix
+
+
+class Trigonometric(StructuralTimeSeries):
+  """Formal representation of a trigonometric seasonal effects model."""
+
+  def __init__(self,
+               period,
+               selected_frequencies,
+               drift_scale_prior=None,
+               initial_effect_prior=None,
+               observed_time_series=None,
+               name=None):
+    """Specify a trigonometric seasonal effects model."""
+
+    with tf.compat.v1.name_scope(
+        name, 'Trigonometric', values=[observed_time_series]) as name:
+
+      _, observed_stddev, observed_initial = (
+          sts_util.empirical_statistics(observed_time_series)
+          if observed_time_series is not None else (0., 1., 0.))
+
+      latent_size = 2 * len(selected_frequencies)
+
+      # Heuristic default priors. Overriding these may dramatically
+      # change inference performance and results.
+      if drift_scale_prior is None:
+        drift_scale_prior = tfd.LogNormal(
+            loc=tf.math.log(.01 * observed_stddev), scale=3.)
+
+      if initial_effect_prior is None:
+        initial_effect_prior = tfd.Normal(
+            loc=observed_initial,
+            scale=tf.abs(observed_initial) + observed_stddev)
+
+      if isinstance(initial_effect_prior, tfd.Normal):
+        initial_state_prior = tfd.MultivariateNormalDiag(
+            loc=tf.stack([initial_effect_prior.mean()] * latent_size, axis=-1),
+            scale_diag=tf.stack([initial_effect_prior.stddev()] * latent_size,
+                                axis=-1))
+      else:
+        initial_state_prior = initial_effect_prior
```

Comment:
Ah yes - I thought I was being clever copying over some functionality from the `Seasonal` component but I have really messed it up - sorry about that.

I've taken your former approach (just exposing `initial_state_prior`) in ed8a438aa2e4db30d6b7ac5bc3811304b25f9932.

That being said, I'm not really sure what is a good default prior for this so would love some advice or references on where e.g. the `Seasonal` default priors come from or if what I have copied over for this default prior is at all sensible.

Reply:
Insofar as we have a philosophy around default priors, it's essentially 'do no harm': we aim for weak priors that are nearly uniform. The whole idea of setting priors based on the observed time series is sort of inherently sinful, so we try to avoid going too far: at best we can pin down plausible values within a few orders of magnitude just to help prevent numerical issues and guide inference away from degenerate solutions. It'd be great to do a rigorous calibration of these prior heuristics someday, but right now it's pretty much seat-of-the-pants. :-)

The default initial state prior in Seasonal (and most other components) is just a zero-mean Gaussian with scale =`abs(observed_initial) + observed_stddev` constructed so that it includes the initial step of the time series and has the ability to cover all the variation in the time series. This makes the conservative assumption that the time series is fully explained by the current component, and in fact by each latent dimension of the current component. In reality there's usually a sum of dimensions/components, and variance compounds over time, so the actual induced prior on the observed series is much wider and weaker; our working hypothesis is that this is fine because we're erring on the side of creating weak priors. By that logic the prior you've copied here seems fine: the scale will be slightly different and in effect wider than that of Seasonal (because this component sums multiple latent dimensions), but as long as we don't promise the semantics of `initial_effect` there's nothing intrinsically wrong with that.